### PR TITLE
Remote file protocol

### PR DESCRIPTION
**Pull request by willmcgugan**
Implements the remote file protocol on agent

### Review
- [x] Ready for review

### CODE REVIEW

File: dataplicity/remote_directory.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    # type: (Text) -> bool
+    """Check a path has no backrefs that may end up serving outside the remote directory root.
+
+    Args:
+        path (Text): A path.
+
+    Returns:
+        bool: True if the path is valid, otherwise False
+    """
+    for component in path.split(os.sep):
```
Comment (toudi) [inline dataplicity/remote_directory.py:53]: you could also try:

```
return os.path.normpath(path).startswith(remote_directory_root)
```

File: dataplicity/remote_directory.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        try:
+            with open(snapshot_path, "rb") as snapshot_file:
+                snapshot_file.seek(offset)
+                data = snapshot_file.read(size)
+                log.debug("read %i bytes from %s", len(data), snapshot_path)
+        except Exception as error:
+            # Most likely, the snapshot file has been removed
+            log.error("failed to read_upload; %r", error)
+            raise ReadFail("unable to read %r" % upload_id)
+        return data
```
Comment (toudi) [inline dataplicity/remote_directory.py:172]: the `return data` could as well be in line 167/168, right ?

**Review by toudi — APPROVED** [review]

File: dataplicity/remote_directory.py (hunk 2 above)
Comment (willmcgugan) [reply dataplicity/remote_directory.py:172]: Could do. I like to put the return at the end wherever possible, so you don't need to go looking for it.

File: dataplicity/remote_directory.py (hunk 1 above)
Comment (willmcgugan) [reply dataplicity/remote_directory.py:53]: Yeah, that probably quicker too.

**Pull request merged**